### PR TITLE
feat: adding LocalRepositoryRegistry

### PR DESCRIPTION
**Pull request by axel7083**
### What does this PR do?

Simply add a store containing the information of where are stored the checkout repository. This store will be required to be able to have https://github.com/projectatomic/ai-studio/issues/27

> This store could be used in the future for https://github.com/projectatomic/ai-studio/issues/234

### What issues does this PR fix or reference?

Fixes https://github.com/projectatomic/ai-studio/issues/250

### How to test this PR?

- [x] Unit tests

### CODE REVIEW

**Comment by lstocchi** [comment]
I'm trying to figure out which is the final result for this. I looked at https://github.com/projectatomic/ai-studio/issues/250 but i didn't get the purpose. Let's say you know which recipe repos the user downloaded, what are you going to do with that info? Is there any new page/icon/message you're going to show?

File: packages/backend/src/registries/LocalRepositoryRegistry.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * The LocalRepositoryRegistry is responsible for keeping track of the directories where recipe are cloned
+ */
+export class LocalRepositoryRegistry {
+  // Map path => LocalRepository
+  private repositories: Map<string, LocalRepository> = new Map();
+
+  constructor(private webview: Webview) {}
+
+  register(localRepository: LocalRepository): podmanDesktopApi.Disposable {
+    this.repositories.set(localRepository.path, localRepository);
```
[the remark this answers is not in the document]
Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:33]: I feel like it is easy to have multiple version of the same recipeId. Having the path as key ensure uniqueness, what if the user decide to clone twice the same recipe in different directory (two different branch let's say)

File: packages/backend/src/registries/LocalRepositoryRegistry.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    this.notify();
+
+    return podmanDesktopApi.Disposable.create(() => {
+      this.unregister(localRepository.path);
+    });
+  }
+
+  unregister(path: string): void {
```
[the remark this answers is not in the document]
Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:41]: This is a registry not a manager, we would probably in the future have a dedicated Manager watching for deletion, but this is not the scope of this PR

File: packages/frontend/src/stores/localRepositories.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import type { Readable } from 'svelte/store';
+import { readable } from 'svelte/store';
+import { MSG_LOCAL_REPOSITORY_UPDATE } from '@shared/Messages';
+import { rpcBrowser, studioClient } from '/@/utils/client';
+import type { LocalRepository } from '@shared/src/models/ILocalRepository';
+
+export const localRepositories: Readable<LocalRepository[]> = readable<LocalRepository[]>([], set => {
```
[the remark this answers is not in the document]
Comment (axel7083) [reply packages/frontend/src/stores/localRepositories.ts]: Oh yeah I forgot about it!! Fixed

File: packages/backend/src/registries/LocalRepositoryRegistry.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:33]: I see your point, but if you select a local repository which has no recipe id ? 

For example, you are creating yourself a repository, and want to open in inside the IA-Studio, at some point you will only have a `ai-studio.yaml` and it has no concept of recipe here.

See https://github.com/redhat-et/locallm/blob/main/chatbot/ai-studio.yaml

[the remark this answers is not in the document]
Comment (lstocchi) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:33]: For me, if you import a local/remote repo like the one mentioned (so the chatbot), it is going to create a new item in the recipe catalog page (maybe under a local section). By looking at the ai-studio.yaml you have all infos to make it work.
The model backend tells you which are the model type supported and we can use that one to fill the models tab. Then to run the application we have the ai-studio.yaml file. 

BTW this is just the way i see it.
@jeffmaury WDYT? Any idea how it should work?

[the remark this answers is not in the document]
Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:33]: I agree we could show the local projects/repository manually added by the user to the recipe page, under as mention some `local` sections.

But they are not recipes, they are a local project/folder/repository, and what's unique about them, their path.

[the remark this answers is not in the document]
Comment (lstocchi) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:33]: You could use some fake id to differentiate them and use the same logic to keep 1-1 recipe-source.

[the remark this answers is not in the document]
Comment (lstocchi) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:33]: It's really a matter if we want to store multiple sources per recipe or not.

[the remark this answers is not in the document]
Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:33]: Why would we generate some fake ids, when we could have the path as id ?

_at some point_ we will have to save the content of the `LocalRepositoryRegistry` to a file, as the projects/folder/repositories imported by the user would have to be still their after a reboot, I would prefer avoiding generating some random values to use that as id, when we could avoid

[the remark this answers is not in the document]
Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:33]: > It's really a matter if we want to store multiple sources per recipe or not.

i would argue no, as you mention, this is probably not something expected, I am more worried about the behavior to adopt for external folder/projects/repositories

File: packages/backend/src/registries/LocalRepositoryRegistry.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * The LocalRepositoryRegistry is responsible for keeping track of the directories where recipe are cloned
+ */
+export class LocalRepositoryRegistry {
+  // Map path => LocalRepository
+  private repositories: Map<string, LocalRepository> = new Map();
+
+  constructor(private webview: Webview) {}
+
+  register(localRepository: LocalRepository): Disposable {
```
Comment (feloy) [inline packages/backend/src/registries/LocalRepositoryRegistry.ts:31]: This registry will be available the time you are in the same instance of the AI Studio app. But how will you build it again when you restart the app (either Podman Desktop or the extension?). Will you have the pair path/recipeId somewhere to be able to rebuild this "database"?

Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:31]: Yes this is the goal _in the future_. For now we do not support adding manually folder/projects/repositories provided by the user. But we would at some point, and we want a place to store that information see https://github.com/projectatomic/ai-studio/issues/234

Today this is simply in memory, but yes in the future we would have to store that information somewhere on disk. But this is not part of this PR

Comment (feloy) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:31]: The solution for the models on disk is to use the modelId as directory name and to rely on it, so we can rebuild this database by scanning the directories in the dedicated directory. Could it be applied here also?

If we don't want to handle the manually created repositories for now, we will have only one repository on disk for each recipe, right? Even if I switch the model, if I understand correctly.

Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:31]: > The solution for the models on disk is to use the modelId

I am not talking about the models id here, but folders containing `ai-studio.yaml` file

> If we don't want to handle the manually created repositories for now, we will have only one repository on disk for each recipe, right? Even if I switch the model, if I understand correctly.

Yes exactly, for now we will only have the cloned repository inside the LocalRepositoryRegistry

Comment (feloy) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:31]: > I am not talking about the models id here, but folders containing ai-studio.yaml file

Yes, I understand, but for me the problem of getting the path of the model's directory is the same as getting the recipy/repository's directory. The reason why I'm comparing both

Comment (axel7083) [reply packages/backend/src/registries/LocalRepositoryRegistry.ts:31]: Yes sure

**Review by feloy — APPROVED** [review]
lgtm

**Pull request merged**